Tests/NIOHTTP1Tests/HTTPServerUpgradeTests.swift

```diff
@@ -2066,5 +2073,31 @@ final class TypedHTTPServerUpgradeTestCase: HTTPServerUpgradeTestCase {
         // We also want to confirm that the upgrade handler is no longer in the pipeline.
         try connectedServer.pipeline.waitForUpgraderToBeRemoved()
     }
+
+    func testHalfClosure() throws {
```

Comment:
Can we add one or two more tests cases here:
1. That sends the request head then input close
2. That sends a full request head & end and then input close to check that we continue the upgrade

Reply:
@FranzBusch 
I did 2

I'm not sure what the expected result is for 1
First I had to send a request head that included a content-length so a `.head` would be passed down the pipeline, but not an`.end` eg
```
OPTIONS * HTTP/1.1\r\nHost: localhost\r\ncontent-length: 10\r\nUpgrade: myproto\r\nKafkaesque: yup\r\nConnection: upgrade\r\nConnection: kafkaesque\r\n\r\n
```
With this it gets stuck in the state `.upgradeReady` because nothing else happens after inputClosed to forward the state machine onwards and the connection stays open. If I return `.close` from the state machine everything works but you explicitly said I shouldn't do that.

Reply:
@FranzBusch any thoughts